libraries/AP_Baro/AP_Baro_BMP085.cpp

```diff
@@ -52,36 +74,99 @@ AP_Baro_BMP085::AP_Baro_BMP085(AP_Baro &baro, AP_HAL::OwnPtr<AP_HAL::I2CDevice>
         _eoc->mode(HAL_GPIO_INPUT);
     }
 
+
+    uint8_t id;
+    
+    if (!_dev->read_registers(0xD0, &id, 1)) {
+        sem->give();
+        return false;
+    }
+    
+    if(id!=0x55) return false; // not BMP180 
+
+
+    _dev->read_registers(0xD1, &_vers, 1);
+
+    bool prom_ok=false;
+    _type=0;
+
     // We read the calibration data registers
-    if (!_dev->read_registers(0xAA, buff, sizeof(buff))) {
-        AP_HAL::panic("BMP085: bad calibration registers");
-    }
-
-    ac1 = ((int16_t)buff[0] << 8) | buff[1];
-    ac2 = ((int16_t)buff[2] << 8) | buff[3];
-    ac3 = ((int16_t)buff[4] << 8) | buff[5];
-    ac4 = ((int16_t)buff[6] << 8) | buff[7];
-    ac5 = ((int16_t)buff[8] << 8) | buff[9];
-    ac6 = ((int16_t)buff[10] << 8) | buff[11];
-    b1 = ((int16_t)buff[12] << 8) | buff[13];
-    b2 = ((int16_t)buff[14] << 8) | buff[15];
-    mb = ((int16_t)buff[16] << 8) | buff[17];
-    mc = ((int16_t)buff[18] << 8) | buff[19];
-    md = ((int16_t)buff[20] << 8) | buff[21];
+    if (_dev->read_registers(0xAA, bb.buff, sizeof(bb.buff))) {
+        prom_ok=true;
+        
+    }
+
+    if(!prom_ok){
+        if(_read_prom((uint16_t *)&bb.buff[0]))  { // BMP180 requires reads by 2 bytes
+            prom_ok=true;
+            _type=1;
+        }
+    }    
+    if(!prom_ok){
+        sem->give();
+        return false;
+    }
+
+    ac1 = ((int16_t)bb.buff[0] << 8) | bb.buff[1];
+    ac2 = ((int16_t)bb.buff[2] << 8) | bb.buff[3];
+    ac3 = ((int16_t)bb.buff[4] << 8) | bb.buff[5];
+    ac4 = ((int16_t)bb.buff[6] << 8) | bb.buff[7];
+    ac5 = ((int16_t)bb.buff[8] << 8) | bb.buff[9];
+    ac6 = ((int16_t)bb.buff[10] << 8) | bb.buff[11];
+    b1 = ((int16_t)bb.buff[12] << 8) | bb.buff[13];
+    b2 = ((int16_t)bb.buff[14] << 8) | bb.buff[15];
+    mb = ((int16_t)bb.buff[16] << 8) | bb.buff[17];
+    mc = ((int16_t)bb.buff[18] << 8) | bb.buff[19];
+    md = ((int16_t)bb.buff[20] << 8) | bb.buff[21];
+    
+    if( (ac1==0 || ac1==-1) ||
+        (ac2==0 || ac2==-1) ||
+        (ac3==0 || ac3==-1) ||
+        (ac4==0 || ac4==0xFFFF) ||
+        (ac5==0 || ac5==0xFFFF) ||
+        (ac6==0 || ac6==0xFFFF) ) return false;
 
     _last_press_read_command_time = 0;
     _last_temp_read_command_time = 0;
 
-    _instance = _frontend.register_sensor();
-
     // Send a command to read temperature
     _cmd_read_temp();
 
     _state = 0;
 
+    _instance = _frontend.register_sensor();
+
     sem->give();
 
     _dev->register_periodic_callback(20000, FUNCTOR_BIND_MEMBER(&AP_Baro_BMP085::_timer, void));
+    return true;
+}
+
+uint16_t AP_Baro_BMP085::_read_prom_word(uint8_t word)
+{
+    const uint8_t reg = 0xAA + (word << 1);
+    uint8_t val[2];
+    if (!_dev->transfer(&reg, 1, val, sizeof(val))) {
+        return 0;
+    }
+    return (val[0] << 8) | val[1];
```

Comment:
doesn't this need to be casted to uint16_t before the shift?

Reply:
no, there's the implicit integer promotion.... the same reason why `f = 1 + 1.1` doesn't make you need to cast it explicitly.

It may be a problem (due to a different reason) if you were running this in a big endian machine, but I don't think we have any supported platform like that.  Anyway, unrelated to this patch, it is clearer IMO to have explicit types be16_t, le16_t, etc and use the conversion functions like is done in BMI160 (le16toh/be16toh) because then it communicates the intent (i.e. sensor works in be or le and the machine works in whatever it is supposed to work, 'h').

Reply:
this just works for ARM CPU